src/classes/view/base.vala

```diff
@@ -40,45 +41,234 @@ namespace pdfpc {
+            // the prerender function
+            int* i = null;
+            // The page_count will be transfered into the lamda function as
+            // well.
+            var page_count = this.get_renderer().metadata.get_slide_count();
+
+            this.prerendering_started();
+
+            Idle.add(() => {
+                if (i == null) {
+                    i = malloc(sizeof(int));
+                    *i = 0;
+                }
+
+                // We do not care about the result, as the
+                // rendering function stores the rendered
+                // pixmap in the cache if it is enabled. This
+                // is exactly what we want.
+                try {
+                    this.get_renderer().render_to_surface(*i);
+                } catch(Renderer.RenderError e) {
+                    error("Could not render page '%i' while pre-rendering: %s", *i, e.message);
+                }
+
+                // Inform possible observers about the cached slide
+                this.slide_prerendered();
+
+                // Increment one slide for each call and stop the loop if we
+                // have reached the last slide
+                *i = *i + 1;
+                if (*i >= page_count) {
+                    this.prerendering_completed();
+                    free(i);
+                    return false;
+                } else {
+                    return true;
+                }
+            });
         }
 
         /**
-         * Display a specific slide
+         * Associate a new Behaviour with this View
          *
-         * If the slide number does not exist a RenderError.SLIDE_DOES_NOT_EXIST is thrown
+         * The implementation supports an arbitrary amount of different
+         * behaviours.
          */
-        public abstract void display(int slide_number, bool force_redraw=false)
-            throws Renderer.RenderError;
+        public void associate_behaviour(Behaviour.Base behaviour) {
+            this.behaviours.append(behaviour);
+            try {
+                behaviour.associate(this);
+            } catch(Behaviour.AssociationError e) {
+                error("Behaviour association failure: %s", e.message);
+            }
+        }
 
         /**
-         * Make the screen black. Useful for presentations together with a whiteboard
+         * Display a specific slide number
+         *
+         * If the slide number does not exist a
+         * RenderError.SLIDE_DOES_NOT_EXIST is thrown
          */
-        public abstract void fade_to_black();
+        public void display(int slide_number, bool force_redraw=false)
+            throws Renderer.RenderError {
+
+            // If the slide is out of bounds render the outer most slide on
+            // each side of the document.
+            if (slide_number < 0) {
+                slide_number = 0;
+            }
+            if (slide_number >= this.slide_limit) {
+                slide_number = this.slide_limit - 1;
```

Comment:
I know, you just moved this around, but these lines of code still caught my eye. `this.slide_limit` was initialized above as `this.n_slides + 1`, so this line sets `slide_number` to an invalid rather than the last slide as claimed in the comment. Maybe `this.n_slides` was intended, which would work out correctly. In that case, the `slide_limit` member could be removed entirely.